### DISCUSSION TITLE
mpd.conf with raspdif (spdif) as audio_output in moode audio

### DISCUSSION THREAD

**Discussion by jtauscher** [MusicPlayerDaemon/MPD#1757] [Q&A]
Hi guys,

moode audio works with the MPD and I found this nice project to have a direct SPDIF output with a raspberry **without** any HAT:

[https://github.com/mill1000/raspdif](url)

It works well with the given example e.g.: 
`ffmpeg -i "/mnt/SDCARD/Stereo Test/LRMonoPhase4.flac" -f s16le -acodec pcm_s16le - | sudo raspdif`

However I am not able to change the audio_output in the mpd.conf like this:
```
decoder {
plugin "ffmpeg"
enabled "yes"
}

input {
plugin "curl"
}

resampler {
plugin "soxr"
quality "high"
threads "1"
}

audio_output {
type "alsa"
name "raspdif"
device "raspdif"
}
```
If I have set mpd.conf like above, no errors occur, MPCs plays music, but no signal on SPDIF GPIO (probably it still outputs on HDMI). 

Settings in moodeaudio are fixed, no options here for the output other than HDMI:

![image](https://user-images.githubusercontent.com/45608914/223993774-6bd81a79-a9bd-465c-b91a-5cdabb7883e6.png)

I'm pretty sure I missed something out - any ideas?

**Comment by MaxKellermann** [comment]
If you run an unpatched MPD build and need support, read https://www.musicpd.org/help/
If you run MoodeAudio's patched MPD version, you can't get support here; ask the MoodeAudio people for help.

**Reply by jtauscher** [reply]
Thanks, I didnt even know there is a difference. Do you have a feeling if my "project" would work with unpatched MPD build? I cant find anybody who ever tried direct SPDIF out with MPD.

**Reply by MaxKellermann** [reply]
I have been using direct SPDIF with MPD for 15+ years, and it works very well, always has. If you can't find somebody like me, maybe you didn't look?
But anyway, that doesn't matter - MPD doesn't know or care what physical medium is connected to a certain ALSA device. ALSA provides an abstraction, which MPD uses to play audio to arbitrary devices, without needing to care which way the audio samples take until the sound waves reach your ears.

**Reply by jtauscher** [reply]
Okay I see, but are you using spdif over HAT device? The raspdif project directly outputs spdif at GPIO21. Anyway, I think the problem is, I cannot set this raspdif device as default for ALSA. I can only select hardware cards. Thats strange because asound.conf is set to:
```
pcm.!default {
  type plug
  slave.pcm "raspdif"
}
pcm.raspdif {
  type plug
  slave {
    pcm {
      type file
      file "/tmp/spdif_fifo"
      format "raw"
      slave.pcm null
    }
    rate 44100
    format S16_LE
    channels 2
  }
  hint {
    description "S/PDIF output via raspdif FIFO"
  }
}

```
So I guess the question is how to force Alsa to this raspdif as default.  Its shown correctly in aplay -L

```
root@moode:~# aplay -L
null
    Discard all samples (playback) or generate zero samples (capture)
_audioout
_audioout__
alsaequal
plug_alsaequal
btstream
camilladsp
crossfeed
plug_bs2b
eqfa12p
plug_eqfa12p
invpolarity
trx_send
default
raspdif
    S/PDIF output via raspdif FIFO
hw:CARD=b1,DEV=0
    bcm2835 HDMI 1, bcm2835 HDMI 1
    Direct hardware device without any conversions
plughw:CARD=b1,DEV=0
    bcm2835 HDMI 1, bcm2835 HDMI 1
    Hardware device with all software conversions
sysdefault:CARD=b1
    bcm2835 HDMI 1, bcm2835 HDMI 1
    Default Audio Device
dmix:CARD=b1,DEV=0
    bcm2835 HDMI 1, bcm2835 HDMI 1
    Direct sample mixing device

```

**Reply by MaxKellermann** [reply]
> but are you using spdif over HAT device?

No.

> I cannot set this raspdif device as default for ALSA

This makes no sense. Why do you want to make it the default?

> asound.conf is set to

What the .... is this? You ask ALSA to write audio samples into a file called `/tmp/spdif_fifo`? This is never going to work properly. If this is what the raspdif designers want you to do, get off this product very very quickly and never use it again. This design is bad. Don't use it. This is so bad that I won't support this kind of setup in MPD.

**Comment by jtauscher** [comment]
You are using a raspberry with direct SPDIF out over GPIO? Or HDMI? The project raspdif is the only way I could find to manage this to output spdif over gpio. The Problem I have my speakers only have ARC/Coac/Opt as digital inputs and I wanted to go with a raspi zero but HDMI is not working (because ARC only). What would be your suggestion to connect digital without any additional hardware?

**Reply by MaxKellermann** [reply]
My main computer has a USB/SPDIF adapter, and in my living room, a Raspi is connected to the amp with HDMI.
